topi/python/topi/image/resize.py

```diff
@@ -210,3 +210,167 @@ def _bicubic(*indices):
+    else:
+        in_n, in_c, in_d, in_h, in_w, in_cc = data.shape
+        output_shape = [in_n, in_c, size[0], size[1], size[2], in_cc]
+
+    if coordinate_transformation_mode == "align_corners":
+        z_ratio = (in_d - 1).astype('float') / (size[0] - 1)
+        y_ratio = (in_h - 1).astype('float') / (size[1] - 1)
+        x_ratio = (in_w - 1).astype('float') / (size[2] - 1)
+    elif coordinate_transformation_mode in ["asymmetric", "half_pixel"]:
+        z_ratio = (in_d).astype('float') / (size[0])
+        y_ratio = (in_h).astype('float') / (size[1])
+        x_ratio = (in_w).astype('float') / (size[2])
+    else:
+        raise ValueError("Unsupported coordinate_transformation_mode: {}".format(
+            coordinate_transformation_mode))
+
+    def _get_pixel(n, c, z, y, x, cc):
+        z = tvm.max(tvm.min(z, in_d - 1), 0)
+        y = tvm.max(tvm.min(y, in_h - 1), 0)
+        x = tvm.max(tvm.min(x, in_w - 1), 0)
+        if layout == 'NDHWC':
+            return data(n, z, y, x, c).astype('float')
+        if layout == 'NCDHW':
+            return data(n, c, z, y, x).astype('float')
+        # else must be NCDHWxc
+        return data(n, c, z, y, x, cc).astype('float')
+
+    def _get_indices(*indices):
+        if layout == 'NDHWC':
+            n, z, y, x, c = indices
+            cc = None
+        elif layout == 'NCDHW':
+            n, c, z, y, x = indices
+            cc = None
+        else:
+            n, c, z, y, x, cc = indices
+
+        return n, c, z, y, x, cc
+
+    def _cast_output(value):
+        if out_dtype:
+            dtype = out_dtype
+        else:
+            dtype = data.dtype
+        return value.astype(dtype)
+
+    # Nearest neighbor computation
+    def _nearest_neighbor(*indices):
+        n, c, z, y, x, cc = _get_indices(*indices)
+
+        in_z = z_ratio * z
+        in_y = y_ratio * y
+        in_x = x_ratio * x
+
+        if coordinate_transformation_mode == "align_corners":
+            zint = tvm.round(in_z).astype('int32')
+            yint = tvm.round(in_y).astype('int32')
+            xint = tvm.round(in_x).astype('int32')
+        elif coordinate_transformation_mode in ["asymmetric", "half_pixel"]:
+            # Add epsilon to floor to prevent gpu rounding errors.
+            epsilon = 1e-5
+            zint = tvm.floor(in_z + epsilon).astype('int32')
+            yint = tvm.floor(in_y + epsilon).astype('int32')
+            xint = tvm.floor(in_x + epsilon).astype('int32')
+        else:
+            raise ValueError("Unsupported coordinate_transformation_mode: {}".format(
+                coordinate_transformation_mode))
+
+        return _cast_output(_get_pixel(n, c, zint, yint, xint, cc))
+
+    # Trilinear helper functions and computation.
+    def _lerp(A, B, t):
+        return A * (1.0 - t) + B * t
+
+    def _trilinear(*indices):
+        n, c, z, y, x, cc = _get_indices(*indices)
+
+        in_z = z_ratio * z
+        in_y = y_ratio * y
+        in_x = x_ratio * x
+
```

Comment:
you need a special handling for half_pixel mode here. See my #4536 and the ONNX spec.

Reply:
Shall I also need apply those changes in "topi/python/topi/testing/bilinear_resize_python.py" and add test cases accordingly?

Reply:
bilinear_resize_python will be updated with my PR, you don't need to touch it. Update trilinear_resize3d_python.py and add test cases for half_pixel (again, see my PR for 2D case)

Reply:
actually in my PR, only resize has coordinate_transform_mode option, upsample still has align_corners and translates align_corners to coordinate_transform_mode of resize (that's what I meant in https://github.com/apache/incubator-tvm/pull/4584#issuecomment-569029721). You updated Upsampling3DAttr too, but that is fine.